quilt/tools/build.py

```diff
@@ -177,49 +178,66 @@ def splitext_no_dot(filename):
     directory tree.
     """
-    def _generate_contents(dir_path):
-        contents = {}
+    def _ignored_name(name):
+        return (
+            name.startswith('.') or
+            name == PACKAGE_DIR_NAME or
+            name.endswith('~') or
+            name == outfilename
+        )
 
+    def _generate_contents(dir_path):
+        safename_duplicates = defaultdict(list)
         for name in os.listdir(dir_path):
-            if name.startswith('.') or \
-               name == PACKAGE_DIR_NAME or \
-               name.endswith('~') or \
-               name == outfilename:
+            if _ignored_name(name):
                 continue
 
             path = os.path.join(dir_path, name)
 
             if os.path.isdir(path):
                 nodename = name
                 ext = None
-                data = _generate_contents(path)
             elif os.path.isfile(path):
                 nodename, ext = splitext_no_dot(name)
-                rel_path = os.path.relpath(path, startpath)
-                data = dict(file=rel_path)
             else:
                 continue
 
-            try:
-                safename = _pythonize_name(nodename)
-                if safename in contents:
-                    if ext:
-                        safename = "{name}_{ext}".format(name=safename, ext=ext)
-                    else:
-                        print("Warning: duplicate name %r in %s." % (safename, dir_path))
-                        continue
-                contents[safename] = data
-            except BuildException:
-                warning = "Warning: could not determine a Python-legal name for {path}; skipping."
-                print(warning.format(path=path))
+            safename = _pythonize_name(nodename)
+            safename_duplicates[safename].append((name, nodename, ext))
+
+        safename_to_name = {}
+        for safename, duplicates in iteritems(safename_duplicates):
+            for name, nodename, ext in duplicates:
+                if (len(duplicates) > 1 or safename in safename_to_name) and ext:
+                    new_safename = _pythonize_name(name)  # Name with ext
+                else:
+                    new_safename = safename
+                existing_name = safename_to_name.get(new_safename)
+                if existing_name is not None:
+                    raise BuildException(
+                        "Duplicate node names. %r was renamed to %r, which overlaps with %r" % (
+                            name, new_safename, existing_name
+                    ))
+                safename_to_name[new_safename] = name
+
+        contents = {}
+        for safename, name in iteritems(safename_to_name):
+            path = os.path.join(dir_path, name)
+
+            if os.path.isdir(path):
+                data = _generate_contents(path)
+            else:
+                rel_path = os.path.relpath(path, startpath)
+                data = dict(file=rel_path)
+
+            contents[safename] = data
```

Comment:
Definitely need unit tests; even if there's existing coverage, it doesn't cover the bugs I'm fixing here.